packages/blocks/src/store/process-block-type.js

```diff
@@ -102,6 +124,9 @@ export const processBlockType =
 			),
 		};
 
+		// Stabilize any experimental supports before applying filters.
+		blockType.supports = stabilizeSupports( blockType.supports );
+
 		const settings = applyFilters(
```

Comment:
What about filters that update `supports` and inject their entries, which could be experimental? What about plugins that expect experimental `support` entries and make decisions based on them?

Reply:
This was raised [earlier](https://github.com/WordPress/gutenberg/pull/63401#discussion_r1696202846) and I believe what prompted @andrewserong to [ask for additional opinions](https://github.com/WordPress/gutenberg/pull/63401#issuecomment-2257635792) on the best path forward.

In a nutshell, the scenario we flagged, requires the migration to occur after the JS filters have been applied. Andrew found however that the block support filters adding attributes might be cleaner, and a better example, if the migration occurred before.

To cover both bases we may need to either
- migrate both, before, and after, filters
- keep the block support filters checking both the experimental and stable flags, then migrate only once after filters are applied

Reply:
It’s also possible that 3rd party plugins run some checks based on the experimental block support syntax so the more I think about it the more it feels you should keep both for some time to let devs adjust or even forever. Best it would be to verify with some plugins that reference these experimental names.

Reply:
Would it be technically feasible to convert the config objects in block supports that have experimental flags into proxy objects and handle the backward compatibility through augmented setters and getters? Let me share a similar case in PHP to better illustrate the idea:

Reply:
To be honest, I'm not sure how far we should take this. While I think folks are probably registering blocks with these experimental supports. I'm not sure people are "reading" these support keys themselves. It feels like not something very useful. This is why I suggested that maybe the current approach is enough and the only real way to know is to try it.

Reply:
Again, I’m perfectly fine with trying as is in this PR to see what issues might pop up and react accordingly. As long is it’s communicated with plugin and theme authors early so they know how to migrate code, where to raise issues or seek for help 👍

Next time, it's going to be easier to repeat the same process based on experience gained this time.

Reply:
Good points! Yeah, I'm leaning toward going with something like the current state of this PR, and essentially committing to allowing plugins to opt-in to block supports, but not explicitly deal with the situation of plugins "reading" the support keys.

In terms of what it'll mean for plugins in the current state:

1. 👍 They can still register blocks and opt-in to typography supports using the `__experimental` prefix.
2. 👍 They can still filter `register_block_type_args` in PHP to opt existing blocks into typography supports that they might not have yet, as the filter in this PR will run last.
3. 👎 They won't be able to filter in JS on `blocks.registerBlockType` to opt blocks in to `__experimental` prefix supports that they don't already have, as the stabilization in JS happens prior to applying the filter.

In order to deal with the potential of `#3`, do you think it'd be worth us running the stabilization a second time after the `blocks.registerBlockType` filter, so that if anything has been opted into, it'll get moved over to the stable key again?

It does mean running the stabilization twice instead of once for each call to process the block type. But it isn't a hugely complex function, so might not be too bad 🤔

I'll this PR as is over the weekend and revisit it sometime next week. Thanks again for all the discussion!

Reply:
One final aside.

> 👎 They won't be able to filter in JS on blocks.registerBlockType to opt blocks in to __experimental prefix supports that they don't already have, as the stabilization in JS happens prior to applying the filter.

I have no real data to support my assumption but based on feedback I got at WordCamps and from GitHub reports, the most common scenario is removing UI controls for these features. I don’t know how that translates into code here though.

Reply:
Removing features is definitely a common scenario. It might be good to know whether folks doing that are not impacted that much.

Reply:
Good points, I think at the very least this PR should handle the basic ways that plugins might disable controls in both JS and PHP. I think this leans me toward applying the transformation twice in JS (once before running the filter, once after), so that we can capture it. Example test code:

```js
function removeTypographySupports( settings, name ) {
	if ( name === 'core/paragraph' && settings.supports.typography ) {
		settings.supports.typography.__experimentalTextTransform = false;
	}
	return settings;
}

addFilter(
	'blocks.registerBlockType',
	'core/style/addAttribute',
	removeTypographySupports
);
```

In `trunk` that'll remove the Letter Case option from the Paragraph block, but not with this PR applied. Ideally it'd recognise if the value has changed after the above filter is run, and migrate it accordingly.

I'll have a play with it this week.

Reply:
>  I think this leans me toward applying the transformation twice in JS (once before running the filter, once after), so that we can capture it.

This would be fantastic, if possible. We have a lot of "Editor curation" content out there that shows folks how to disable/enable block supports with JS and PHP. Either way, I think the developer community will be quite excited about this stabilization effort. 

Even if the transformation was not applied twice, you could do something like this to support 6.7 while maintaining backward compatibility, right? 

```
function removeTypographySupports( settings, name ) {
	if ( name === 'core/paragraph' && settings.supports.typography ) {
		settings.supports.typography.__experimentalTextTransform = false;
		settings.supports.typography.textTransform = false;
	}
	return settings;
}

addFilter(
	'blocks.registerBlockType',
	'core/style/addAttribute',
	removeTypographySupports
);
```

Reply:
> Even if the transformation was not applied twice, you could do something like this to support 6.7 while maintaining backward compatibility, right?

Yes, that example should work nicely for plugins that wish to be forward and backward compatible.

For now, I _think_ I've gotten it working fairly well by applying the transformation a second time for both the block supports and deprecations in https://github.com/WordPress/gutenberg/pull/63401/commits/6934b8aa39dc7fecb95a8a066c8fed5c7bc544d0. I've added a couple of tests to cover the scenario we've described here.

The implementation still assumes that plugins aren't _watching_ for current values before overriding things, but at least this gives us coverage for plugins that are attempting to switch things off.

I'll give this PR a bit more of a manual test, though, as it _seems_ right to me now, but I'm not sure if I'm thinking is right 😄